R/SEIRmodels.R

```diff
@@ -0,0 +1,749 @@
+#' @title infections for SIR model
+#' @param I I
+#'
+#' @param S S
+#' @param pars list
+#' @export
+#'
+#' @importFrom stats rbinom
+Infections <- function(I, S, pars) {
+
+  # SIR: two types of events for S, so competing hazards. A fraction of
+  # S events are deaths and the rest are infections.
+
+  FOI <- (pars$beta * I) / pars$N
+  fmu <- FOI + pars$mu
+  fmudt <- fmu * pars$dt
+
+  prob1 <- 1.0 - exp(-1.0 * fmudt)
+  prob2 <- pars$mu / fmu
+
+  n_events_S <- rbinom(1, S, prob1)
+
+  if (n_events_S > 0) {
+    n_deaths_S <- rbinom(1, n_events_S, prob2)
+    n_infections_S <- n_events_S - n_deaths_S
+  }
+  else {
+    n_deaths_S <- 0
+    n_infections_S <- 0
+  }
+
+  list(n_deaths_S = n_deaths_S, n_infections_S = n_infections_S)
+}
+
+Recoveries <- function(I, pars) {
+  # SIR: two types of events for I, so competing hazards and a fraction of
+  # I events are deaths and the rest are recoveries
+  coeff <- pars$nu + pars$mu
+  coeffdt <- coeff * pars$dt
+
+  prob1 <- 1.0 - exp(-1.0 * coeffdt)
+  prob2 <- 1.0 - exp(-1.0 * pars$mu/coeff)
+
+  n_events_I <- rbinom(1, I, prob1)
+
+  n_deaths_I <- rbinom(1, n_events_I, prob2)
+  n_recoveries_I <- n_events_I - n_deaths_I
+
+  list(n_deaths_I = n_deaths_I, n_recoveries_I = n_recoveries_I)
+}
+
+#' @importFrom stats rbinom
+Births <- function(R, pars) {
+
+  coeffdt <- pars$mu * pars$dt
+  prob <- 1.0 - exp(-1.0 * coeffdt)
+  n_deaths_R <- rbinom(1, R, prob)
+  n_births <- pars$n_deaths_S  + pars$n_deaths_I + n_deaths_R
+
+  list(n_deaths_R = n_deaths_R, n_births = n_births)
+}
+
+Update <- function(S, I, R, pars) {
+
+  news <- S - pars$n_deaths_S - pars$n_infections_S + pars$n_births
+  newi <- I + pars$n_infections_S  - pars$n_recoveries_I  - pars$n_deaths_I
+  newr <- R + pars$n_recoveries_I - pars$n_deaths_R
+  newN <- news + newi + newr
+
+  list(news = news, newi = newi, newr = newr)
+}
+
+Displaythemodel <- function(df) {
```

Comment:
I can't tell when to use display1, 2 or 3.... Could we have one function? Or rename/document each of them